src/mono/wasm/debugger/DebuggerTestSuite/BreakpointTests.cs

```diff
@@ -0,0 +1,228 @@
+                columnNumber = 2,
+                url = "dotnet://debugger-test.dll/this-file-doesnt-exist.cs",
+            });
+
+            var bp1_res = await cli.SendCommand("Debugger.setBreakpointByUrl", bp1_req, token);
+
+            Assert.True(bp1_res.IsOk);
+            Assert.Empty(bp1_res.Value["locations"].Values<object>());
+            //Assert.Equal ((int)MonoErrorCodes.BpNotFound, bp1_res.Error ["code"]?.Value<int> ());
+        }
+
+        [Fact]
+        public async Task CreateGoodBreakpointAndHit()
+        {
+            var bp = await SetBreakpoint("dotnet://debugger-test.dll/debugger-test.cs", 10, 8);
+
+            var eval_req = JObject.FromObject(new
+            {
+                expression = "window.setTimeout(function() { invoke_add(); }, 1);",
+            });
+
+            await EvaluateAndCheck(
+                "window.setTimeout(function() { invoke_add(); }, 1);",
+                "dotnet://debugger-test.dll/debugger-test.cs", 10, 8,
+                "IntAdd",
+                wait_for_event_fn: (pause_location) =>
+                {
+                    Assert.Equal("other", pause_location["reason"]?.Value<string>());
+                    Assert.Equal(bp.Value["breakpointId"]?.ToString(), pause_location["hitBreakpoints"]?[0]?.Value<string>());
+
+                    var top_frame = pause_location["callFrames"][0];
+                    Assert.Equal("IntAdd", top_frame["functionName"].Value<string>());
+                    Assert.Contains("debugger-test.cs", top_frame["url"].Value<string>());
+
+                    CheckLocation("dotnet://debugger-test.dll/debugger-test.cs", 8, 4, scripts, top_frame["functionLocation"]);
+
+                    //now check the scope
+                    var scope = top_frame["scopeChain"][0];
+                    Assert.Equal("local", scope["type"]);
+                    Assert.Equal("IntAdd", scope["name"]);
+
+                    Assert.Equal("object", scope["object"]["type"]);
+                    CheckLocation("dotnet://debugger-test.dll/debugger-test.cs", 8, 4, scripts, scope["startLocation"]);
+                    CheckLocation("dotnet://debugger-test.dll/debugger-test.cs", 14, 4, scripts, scope["endLocation"]);
+                    return Task.CompletedTask;
+                }
+            );
+        }
+
+        [Theory]
+        [InlineData("invoke_add", "IntAdd", "c == 30", true)]
+        [InlineData("invoke_add", "IntAdd", "true", true)]
+        [InlineData("invoke_add", "IntAdd", "5", true)]
+        [InlineData("invoke_add", "IntAdd", "c < 40", true)]
+        [InlineData("invoke_add", "IntAdd", "c == 40", false)]
+        [InlineData("invoke_add", "IntAdd", "g == 40", false)]
+        [InlineData("invoke_add", "IntAdd", "c < 0", false)]
+        [InlineData("invoke_use_complex", "UseComplex", "complex.A == 10", true)]
+        public async Task ConditionalBreakpoint(string function_to_call, string method_to_stop, string condition, bool bp_index_expected)
```

Comment:
I think splitting this into 3 different data sets would be useful. Like:

```csharp
        public static TheoryData<string, string, string, bool> FalseConditions = new TheoryData<string, string, string, bool>
        {
            { "invoke_add", "IntAdd", "c == 40", false },
            { "invoke_add", "IntAdd", "g == 40", false },
            { "invoke_add", "IntAdd", "c < 0", false },
        };

        public static TheoryData<string, string, string, bool> TrueConditions = new TheoryData<string, string, string, bool>
        {
            { "invoke_add", "IntAdd", "c == 30", true },
            { "invoke_add", "IntAdd", "true", true },
            { "invoke_add", "IntAdd", "5", true },
            { "invoke_add", "IntAdd", "c < 40", true },
            { "invoke_use_complex", "UseComplex", "complex.A == 10", true },
        };

        public static TheoryData<string, string, string, bool> InvalidConditions = new TheoryData<string, string, string, bool>
        {
            { "invoke_add", "IntAdd", "foo.bar", false },
            { "invoke_add", "IntAdd", "Math.IntAdd()", false },
            { "invoke_add", "IntAdd", "c == \"xyz\"", false },
            { "invoke_add", "IntAdd", "Math.NonExistantProperty", false },
            { "invoke_add", "IntAdd", "\"hello\"", false },
            { "invoke_add", "IntAdd", "null", false },
        };

        [Theory]
        [MemberData(nameof(FalseConditions))]
        [MemberData(nameof(TrueConditions))]
        [MemberData(nameof(InvalidConditions))]
```

You might need to adjust the results, depending on what we want to do in case of conditions that we couldn't evaluate.

I also noticed that it failed for `condition="\"foo\""`, because 
`if (retValue?["value"]?.Value<bool>() == true)`
.. the `.Value<bool>` here throws. So:

```diff
-                if (retValue?["value"]?.Value<bool>() == true)
-                    return true;
+                var valueStr = retValue?["value"]?.Value<string>();
+                if (!string.IsNullOrEmpty(valueStr) && bool.TryParse(valueStr, out bool evalValue))
+                    return evalValue;
```

The above change made the test with `condition="5"` fail. The test expects that to evaluate to true, and `.Value<bool>` seems to return `true` for that! The above condition will return false.

So, that test might need to be adjusted depending on what we want to do for that, and the other non-bool results.

Reply:
No, the condition=5 is working, and it's returning true as expected.
I have fixed for other types, it will be available in my next commit.

Reply:
Changed to use TheoryData.